optimize/protocol.py

```diff
@@ -0,0 +1,254 @@
+        work_stddev(i -> i+1) <= work_stddev_threshold,
+        work_stddev(i+1 -> i) <= work_stddev_threshold
+        for all i
+
+    Parameters
+    ----------
+    lambdas_k : monotonic sequence starting at 0 and ending at 1
+    f_k : estimated reduced free energies of initial lambda windows
+    u_kn : reduced potential energies of initial samples in all initial lambda windows
+    N_k : number of samples from each initial lambda window
+    work_stddev_threshold : controls spacing of new lambda windows (smaller threshold -> tighter spacing)
+
+    Notes
+    -----
+    Applies the following approximations:
+    * u(x_n, lam) for new trial values of lam can be well-approximated by linear interpolation of u_kn
+    * work_stddev(prev_lam, next_lam) can be well-approximated by reweighting samples from initial protocol
+    """
+    # aggregate all samples from initial protocol
+    reference_log_weights_n = log_weights_from_mixture(u_kn, f_k, N_k)
+
+    # linearly interpolate initial energies
+    _, vec_u_interp, vec_delta_u_interp = linear_u_kn_interpolant(lambdas_k, u_kn)
+
+    # function to estimate work_stddev(prev_lam, next_lam)
+    work_stddev_estimator = construct_work_stddev_estimator(reference_log_weights_n, vec_u_interp, vec_delta_u_interp)
+
+    # function needed to place the next lambda window given the location of the previous window
+    assess_lambda_pair = partial(
+        construct_heuristic_lambda_pair_assessor(work_stddev_estimator),
+        desired_stddev=work_stddev_threshold,
+    )
+
+    # build a new protocol one state at a time
+    optimized_protocol = greedily_optimize_protocol(assess_lambda_pair)
+    return optimized_protocol
+
+
+def log_weights_from_mixture(u_kn: Array, f_k: Array, N_k: Array) -> Array:
+    """Assuming
+    * K energy functions u_k
+    * N_k samples from each state e^{-u_k} / Z_k
+    * free energy estimates f_k ~= - log Z_k
+    * u_kn[k, n] = u_k(xs[n)
+        (where samples xs = concatenation of all samples, in any order,
+        and len(xs) == N = sum(N_k))
+
+    interpret the collection of N = \sum_k N_k samples as coming from a
+    mixture of states p(x) = (1 / K) \sum_k e^-u_k / Z_k
+    """
+    log_q_k = f_k - u_kn.T
+    N_k = np.array(N_k, dtype=np.float64)  # may be ints, or in a list...
+    log_weights = logsumexp(log_q_k, b=N_k, axis=1)
+    return log_weights
+
+
+def linear_u_kn_interpolant(lambdas: Array, u_kn: Array) -> Tuple[Callable, Callable, Callable]:
+    """Given a matrix u_kn[k, n] = u(xs[n], lambdas[k]) produce linear interpolated estimates of u(xs[n], lam)
+    at arbitrary new values lam"""
+
+    def u_interp(u_n: Array, lam: Float) -> Float:
+        return np.nan_to_num(np.interp(lam, lambdas, u_n), nan=+np.inf, posinf=+np.inf)
+
+    @jit
+    def vec_u_interp(lam: Float) -> Array:
+        return vmap(u_interp, (1, None))(u_kn, lam)
+
+    @jit
+    def vec_delta_u(from_lam: Float, to_lam: Float) -> Array:
```

Comment:
Turns out this function can be removed entirely! Addressed in https://github.com/proteneer/timemachine/pull/538/commits/9bae95ba1e14a3d52a42b16642128ce076956ca1 

(Comment: I had added this nan_to_num step when flipping switches to avoid a gradient NaN in an earlier approach. The current approach is gradient-free, and a less unsound workaround should be possible if we later need to estimate gradients of a similar objective.)